Jennyf/up

jennyf19 opened the pull request:
I have not completed the required manual/automated tests for this branch, but wanted to get a PR out for review. I will update the tests as I complete them.

Only changes starting with [commit 05541d8](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/commit/05541d804a45c6b00e2261467806d4ec5e03b6f5) need review, as all previous commits have already been reviewed.

**Testing completed:**
> ADALDesktopTestApp (u/p) - manual
> MSAL DesktopTestApp (u/p) - manual

rayluo committed:
Refactor UserCredential to use a common base class in core

rayluo committed:
A tentative MexParser implementation. We have not changed UnitTests yet.

rayluo committed:
Refactor HttpRequest to accept an HttpContent as body

rayluo committed:
WsTrust prototype

rayluo committed:
Implement UserRealmDiscovery in core, and use it in adal

rayluo committed:
Implement WIA in MSAL

rayluo committed:
Unit tests for WsTrust/*.cs

rayluo committed:
PublicClientApplicationTests.cs

rayluo committed:
Change core and msal resource settings to always-copy

rayluo committed:
One workaround for the DeploymentItem weirdness

rayluo committed:
MSAL follows same test resource file names in ADAL

Otherwise you will hit the nasty DeploymentItem gotcha documented in here
http://www.ademiller.com/blogs/tech/2007/10/gotchas-mstests-deploymentitem-attribute/

rayluo committed:
Rename Core.UserCred to Core.UserCredential

rayluo committed:
Rename UserCred.cs to UserCreential.cs

rayluo committed:
ADAL UserRealmDiscoveryResponse & WSTrust/* retire

rayluo committed:
UIAutomation project needs to include core assembly now

rayluo committed:
Modify MSAL Desktop Test App to test WIA manually

rayluo committed:
Remove contents only needed by future Username Password flow

rayluo committed:
Put an enum into a separate file

rayluo committed:
Make Core.UserCredential internal

rayluo committed:
Refactor MexParser to new Core Exceptions

rayluo committed:
Change WsTrustRequest for Core Exceptions

rayluo committed:
Addressed PR comment: log the unknown user type

rayluo committed:
Changing 2 MsalException to MsalClientException

rayluo committed:
Refactor Exception handling in WsTrustRequest.cs

rayluo committed:
Renaming a _UserCredential to userCredential

rayluo committed:
Remove randomness when finding MEX policies

rayluo committed:
Removed trailing empty lines accidentally added by my editor

rayluo committed:
Rebase to latest project structure and breaking changes

bgavrilMS committed:
Re-enable IWA and U/P on ADAL, IWA on MSAL

bgavrilMS committed:
Fix token view in sample app and computation of realm uri

bgavrilMS committed:
Minor changes to comments

bgavrilMS committed:
Update comments and add conditional compilation

bgavrilMS committed:
Fix missing private method on netcore

bgavrilMS committed:
Ensure User Password in ADAL throws the correct exception when password is wrong

bgavrilMS committed:
wip

bgavrilMS committed:
Ensure User Password in ADAL throws the correct exception when password is wrong (#1255)

jennyf19 committed:
Merge remote-tracking branch 'origin/bogavril/iwa2' into jennyf/up

jennyf19 committed:
Initial commit for u/p flow  - work in progress.

jennyf19 committed:
merge conflicts

jennyf19 committed:
address merge conflicts

jennyf19 committed:
fix build failure

jennyf19 committed:
updates

jennyf19 committed:
add flow to msal desktop app and add oauth params

jennyf19 committed:
add managed user tests
fix up a few issues

MarkZuber committed:
Update infra to support username/password flow

jennyf19 committed:
commenting out 2 u/p tests to create release build...will fix.

jennyf19 committed:
fix federated test cases

jennyf19 committed:
Add unit tests
Add XML comments

bgavrilMS commented on msal/samples/desktop/SampleApp/MainForm.cs:
```diff
@@ -52,31 +56,70 @@ public MainForm()
-                string token = await _msalHelper.GetTokenForCurrentAccountAsync(new[] {"user.read"}, account)
+                string token = await _msalHelper.GetTokenForCurrentAccountAsync(new[] { "user.read" }, account)
                     .ConfigureAwait(false);
-                DisplayUserInformationFromGraph(token);
```
Why was this removed?  #Closed

jennyf19 replied:
it wasn't working...i don't think it ever has worked.

---
In reply to: [219452300](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219452300) [](ancestors = 219452300)

bgavrilMS commented on msal/samples/desktop/SampleApp/MsalAuthHelper.cs:
```diff
@@ -98,5 +112,32 @@ public async Task<string> GetTokenForCurrentAccountAsync(IEnumerable<string> sco
+                result = await Application.AcquireTokenByIntegratedWindowsAuthAsync(scopes, user).ConfigureAwait(false);
+                return result.AccessToken;
+            }
+            catch (MsalServiceException ex)
```
You can just catch all exception, like at line 130 and then show the message box, like at line 137.  #Resolved

bgavrilMS commented on msal/samples/desktop/SampleApp/MsalAuthHelper.cs:
```diff
@@ -37,54 +38,67 @@ namespace SampleApp
+                    result = await Application.AcquireTokenAsync(scopes).ConfigureAwait(false);
+                    return result.AccessToken;
+                }
+                catch (MsalServiceException ex)
```
Not needed since you catch all exceptions at line 74 #Resolved

bgavrilMS commented on msal/src/Microsoft.Identity.Client/ClientApplicationBase.cs:
```diff
@@ -277,6 +277,11 @@ public async Task RemoveAsync(IAccount account)
         internal async Task<AuthenticationResult> AcquireTokenSilentCommonAsync(Authority authority,
             IEnumerable<string> scopes, IAccount account, bool forceRefresh, ApiEvent.ApiIds apiId)
         {
+            if (account == null)
```
This is part of a different change and should not be here. I think you forgot to rebase all changes to the current dev. Still, this should not be here. #Pending

jennyf19 replied:
i had to add this for testing, but I did rebase w/current dev, but maybe because i added it earlier (for testing), the changes w/dev were not accounted for?

---
In reply to: [219454270](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219454270) [](ancestors = 219454270)

bgavrilMS commented on msal/src/Microsoft.Identity.Client/Exceptions/MsalErrorMessage.cs:
```diff
@@ -130,6 +130,8 @@ internal static class MsalErrorMessage
 
         public const string DuplicateQueryParameterTemplate = "Duplicate query parameter '{0}' in extraQueryParameters";
         public const string DeviceCertificateNotFoundTemplate = "Device Certificate was not found for {0}";
+        public const string MsalUiRequiredMessage =
```
Different change #Resolved

bgavrilMS commented on msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs:
```diff
@@ -85,14 +84,22 @@ private async Task FetchAssertionFromWsTrustAsync()
+                    // handle grant flow
+                    if (!this.usernamePasswordInput.HasPassword())
+                    {
+                        throw new MsalException(MsalError.PasswordRequiredForManagedUserError);
```
We should always throw MsalClientException or MsalServiceException. We expose MsalException as a base class as a convenience for users to catch and treat both at the same time, but we should never throw it.  #Resolved

henrik-me replied:
Seems like this should be a client exception as Bogdan suggests. Same is in IWarRequest

---
In reply to: [219454654](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219454654) [](ancestors = 219454654)

bgavrilMS commented on msal/src/Microsoft.Identity.Client/PublicClientApplication.cs:
```diff
@@ -511,16 +511,43 @@ public async Task<AuthenticationResult> AcquireTokenByIntegratedWindowsAuthAsync
+        /// <param name="username"></param>
+        /// <param name="password"></param>
+        /// <returns></returns>
+        public async Task<AuthenticationResult> AcquireTokenByUsernamePasswordAsync(IEnumerable<string> scopes, string username, string password)
```
I would propose that we don't expose this overload (at least not on .net desktop where we have SecureString). The reasoning is that we should offer users a single API for achieving U/P, and we should prefer the more secure version of this API (i.e. the one at line 534 with SecureString). @henrik-me - can you think of a user scenario where they cannot use SecureString but need to use string  on net45 ? #Resolved

henrik-me replied:
I can't think of a scenario right now. but there might be one... 

---
In reply to: [219455652](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219455652) [](ancestors = 219455652)

jennyf19 replied:
@henrikm and i discussed this...we will only expose securestring for now.

---
In reply to: [219528465](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219528465) [](ancestors = 219528465,219455652)

bgavrilMS commented on core/src/Exceptions/CoreErrorMessages.cs:
```diff
@@ -83,7 +83,7 @@ internal class CoreErrorMessages
 
-
-
+        public const string InternalErrorCacheEmptyUsername = "Internal error - trying to remove an ADAL user with an empty username. Possible cache corruption. See https://aka.ms/adal_token_cache_serialization";
```
Merge concflicts ? Are these related to IWA / UP ? #Resolved

jennyf19 replied:
yes

---
In reply to: [219456012](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219456012) [](ancestors = 219456012)

bgavrilMS commented on core/tests/Test.Microsoft.Identity.Core.Unit/Mocks/MockHttpMessageHandler.cs:
```diff
@@ -47,6 +47,8 @@ internal class MockHttpMessageHandler : HttpMessageHandler
 
         public IDictionary<string, string> PostData { get; set; }
 
+        public IDictionary<string, object> PostDataObject { get; set; }
```
Shouldn't this be used  in this class? #Resolved

jennyf19 replied:
this is needed for the mocks, to handle the securestring.

---
In reply to: [219473186](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219473186) [](ancestors = 219473186)

bgavrilMS commented on msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs:
```diff
@@ -122,9 +122,17 @@ private async Task UpdateUsernameAsync()
-            client.AddBodyParameter(OAuth2Parameter.GrantType, OAuth2GrantType.Password);
-            client.AddBodyParameter(OAuth2Parameter.Username, this.usernamePasswordInput.UserName);
-            client.AddBodyParameter(OAuth2Parameter.Password, new string(this.usernamePasswordInput.PasswordToCharArray()));
+            if (this.userAssertion != null)
```
I remember you saying this condition is hit only when the account is managed? If so, could you add a comment to make it clearer? #Resolved

bgavrilMS commented on msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs:
```diff
@@ -134,8 +133,13 @@ protected override void SetAdditionalRequestParameters(OAuth2Client client)
 
-            // To request id_token in response
-            client.AddBodyParameter(OAuth2Parameter.Scope, OAuth2Value.ScopeOpenId);
+            SortedSet<string> unionScope =
```
Why SortedSet instead of a HashSet? Is it for debugging, i.e. to see have the Get params sorted alphabetically in Fiddler? #Resolved

bgavrilMS commented on core/src/OAuth2/OAuthConstants.cs:
```diff
@@ -102,6 +102,8 @@ internal class OAuth2Value
         public static readonly string[] ReservedScopes = {"openid", "profile", "offline_access"};
```
You can update this constant to use ScopeOpenId, ScopeOfflineAccess etc. #Resolved

bgavrilMS commented on msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:
```diff
@@ -415,164 +415,165 @@ public async Task AcquireTokenByIntegratedWindowsAuthTest()
-
-            Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
-        }
+        //[TestMethod]
```
Lots of dead code, I assume these are uncommented in a subsequent commit? #Resolved

bgavrilMS commented on msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:
```diff
@@ -416,164 +416,172 @@ public async Task AcquireTokenByIntegratedWindowsAuthTest()
-
-        //    Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
-        //}
+        [TestMethod]
```
So they are... #Resolved

bgavrilMS commented on msal/src/Microsoft.Identity.Client/PublicClientApplication.cs:
```diff
@@ -512,28 +512,28 @@ public async Task<AuthenticationResult> AcquireTokenByIntegratedWindowsAuthAsync
+        /// <param name="username">Identifier of the user application requests token on behalf.</param>
+        /// <param name="securePassword">User password.</param>
+        /// <returns>Authentication result containing a token for the requested scopes and account</returns>
+        public async Task<AuthenticationResult> AcquireTokenByUsernamePasswordAsync(IEnumerable<string> scopes, string username, System.Security.SecureString securePassword)
```
@jmprieur to review the public API comments.

bgavrilMS commented on msal/tests/Test.MSAL.NET.Unit/IWATests.cs:
```diff
@@ -0,0 +1,623 @@
+        [TestCategory("IWATests")]
+        [DeploymentItem(@"Resources\TestMex.xml", "MsalResource")]
+        [DeploymentItem(@"Resources\WsTrustResponse13.xml", "MsalResource")]
+        public async Task AcquireTokenByIntegratedWindowsAuthTest()
```
Excellent!

jennyf19 replied:
:)

---
In reply to: [219475814](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219475814) [](ancestors = 219475814)

bgavrilMS commented on msal/tests/Test.MSAL.NET.Unit/IWATests.cs:
```diff
@@ -0,0 +1,623 @@
+            });
+
+            // MEX
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
```
There seem to be a lot of these duplicated code blocks for setting up mocks. See if you can create some private methods that can be reused.  #Resolved

jennyf19 replied:
yes...thanks!

---
In reply to: [219476190](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219476190) [](ancestors = 219476190)

bgavrilMS commented on msal/tests/Test.MSAL.NET.Unit/IWATests.cs:
```diff
@@ -0,0 +1,623 @@
+namespace Test.MSAL.NET.Unit
+{
+    [TestClass]
+    public class IWATests
```
There are username password tests in here as well, not just iwa #Resolved

jennyf19 replied:
i renamed it, but open to a better name. 

---
In reply to: [219477471](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219477471) [](ancestors = 219477471)

bgavrilMS commented on msal/tests/Test.MSAL.NET.Unit/IWATests.cs:
```diff
@@ -0,0 +1,623 @@
+
+        [TestMethod]
+        [TestCategory("IWATests")]
+        public async Task ManagedUsernamePasswordAcquireTokenTest()
```
Could we add a test for UsernamePassword where the password in wrong? I think this is a big scenario that should be covered. #Resolved

bgavrilMS reviewed (changes requested):
Looks good, some changes requested.

bgavrilMS committed:
Improve error message when user is not the domain joined user

henrik-me commented on msal/src/Microsoft.Identity.Client/Exceptions/MsalErrorMessage.cs:
```diff
@@ -97,7 +97,7 @@ internal static class MsalErrorMessage
 
         public const string UnknownUser = "Could not identify logged in user";
-        public const string UnsupportedUserType = "Unsupported User Type '{0}'";
+        public const string UnsupportedUserType = "Unsupported User Type '{0}'. Possible cause: the user {1} is not domain joined.";
```
pls. add an aka.ms link to our wiki pages talking about this error. then we can update our docs to explain how to resolve it and even in the service cases point to external docs. As a starting point I suggest simply pointing to an error page for client exceptions and anotherone for service exceptions. happy to discuss more. Clarification: we should imo also add a link to a general error page, though if the one already linked clearly explains the error scenario then that is naturally better (though the general error information page should explain this as well or point to the right place).  #Resolved

henrik-me commented on msal/tests/dev apps/DesktopTestApp/DesktopTestApp.csproj:
```diff
@@ -42,6 +42,7 @@
     <OutputPath>bin\Appveyor\</OutputPath>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Microsoft.VisualBasic" />
```
? #Resolved

bgavrilMS replied:
Yeah, it looks like VB has a simple dialog that can be used to fetch an input box. It helps with testing, although you cannot control the text box to set it to mask the text.  #Resolved

jennyf19 replied:
i removed the reference.

---
In reply to: [219518522](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219518522) [](ancestors = 219518522)

bgavrilMS committed:
Add aka.ms links to error messages

jennyf19 committed:
Address PR comments:
- Use consts for strings (OAuth2Value)
- Add aka.ms link for error message
- UsernamePasswordRequest class:
  - Add comment for managed account
  - Change SortedSet<> to HashSet<>
  - Throw MsalClientExceptions instead of MsalExceptions
- Remove public api of u/p w/string as parameter..have moved to migration aid doc instead.
- Add more unit tests

jennyf19 committed:
merge

jennyf19 commented on msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs:
```diff
@@ -94,8 +94,13 @@ private async Task FetchAssertionFromWsTrustAsync()
+                            CultureInfo.CurrentCulture, 
+                            MsalErrorMessage.UnsupportedUserType, 
+                            userRealmResponse.AccountType, 
+                            this.iwaInput.UserName));
```
Is this considered Pii? #Resolved

henrik-me replied:
yes, however if only shown to the user it is fine. if we log it... we have to filter it out

---
In reply to: [219635839](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219635839) [](ancestors = 219635839)

jennyf19 replied:
ok. we're not logging it. #Resolved

henrik-me commented on msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/IWARequest.cs:
```diff
@@ -0,0 +1,128 @@
+    /// <summary>
+    /// Handles requests that are non-interactive. Currently MSAL supports Integrated Windows Auth.
+    /// </summary>
+    internal class IWARequest : RequestBase
```
>IWARequest [](start = 19, length = 10)

Perhaps this should be spelled out. IWA sounds like some interface... #Resolved

jennyf19 replied:
I spelled it out everywhere.

---
In reply to: [219640027](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219640027) [](ancestors = 219640027)

henrik-me commented on msal/tests/dev apps/DesktopTestApp/MainForm.cs:
```diff
@@ -158,6 +161,45 @@ private async void AcquireTokenInteractive_Click(object sender, EventArgs e)
+            ClearResultPageInfo();
+
+            string username = loginHintTextBox.Text; //Can be blank for U/P
+            var password = Microsoft.VisualBasic.Interaction.InputBox("Password?");
```
> Microsoft.VisualBasic.Interaction.InputBox("Password?"); [](start = 26, length = 57)

nit: It is possible to set the text box in the ui to be a password box... then no need to introduce a VB control? #Resolved

jennyf19 replied:
yes. but less fun.

---
In reply to: [219640245](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219640245) [](ancestors = 219640245)

jennyf19 committed:
updates to sample apps to handle securestring

henrik-me commented on msal/samples/desktop/SampleApp/MainForm.cs:
```diff
@@ -52,31 +56,83 @@ public MainForm()
+        private async void acquireTokenUsernamePasswordButton_Click(object sender, EventArgs e)
+        {
+            tokenResultBox.Text = string.Format("");
+            string password = tokenResultBox.Text;
```
>string password = tokenResultBox.Text; [](start = 12, length = 38)

you should pass the textboxresult to the method, otherwise you are copying the string into memory anyway... here you end up doing it twice #Resolved

henrik-me replied:
then...:

var secure = new SecureString();
foreach (char c in textbox1.Text) {
  secure.AppendChar(c);
}

---
In reply to: [219649112](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219649112) [](ancestors = 219649112)

jennyf19 replied:
defeats the point of the secure string...i fixed this.

---
In reply to: [219649259](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219649259) [](ancestors = 219649259,219649112)

jennyf19 committed:
Address PR Comments:
- Rename a few things from IWA to longer version
- Fix handling of secure string in sample apps
- Add unit tests for common authority
- add invalid request response for mocks

jennyf19 committed:
remove VB reference from desktop app

jennyf19 committed:
uap test app issues

jennyf19 committed:
remove uap test app...bogdan will add it in again later.

bgavrilMS commented on msal/src/Microsoft.Identity.Client/MigrationAid/MigrationAid.cs:
```diff
@@ -145,4 +146,26 @@ public partial class TokenCacheNotificationArgs
+    /// Abstract class containing common API methods and properties. 
+    /// For details see https://aka.ms/msal-net-client-applications
+    /// </Summary>
+    public partial class PublicClientApplication
```
>PublicClientApplication [](start = 25, length = 23)

We never had this method in MSAL, why do we want to obselete an inexistent method? #Resolved

henrik-me replied:
This is to help those upgrading from ADAL trying to use the string version of the pwd. Telling them to use the SecureString intead.
